### PR TITLE
regen certs on update

### PR DESCRIPTION
**Pull request by mjudeikis**
Regenerates certificate if dependency property changed.

### CODE REVIEW

File: pkg/api/plugin.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type Plugin interface {
 
 	// GenerateConfig ensures all the necessary in-cluster config is generated
 	// for an Openshift cluster.
-	GenerateConfig(ctx context.Context, cs *OpenShiftManagedCluster) error
+	GenerateConfig(ctx context.Context, cs, oldCS *OpenShiftManagedCluster) error
```
Comment (jim-minter) [inline pkg/api/plugin.go]: doing this is unfortunately out of the question.  oldCS is not available on the RP backend.  You don't need it anyway.

File: pkg/config/generate.go [hunk 2]
```diff
@@ -227,11 +230,14 @@ func Generate(cs *api.OpenShiftManagedCluster) (err error) {
...
-		if *cert.key != nil && *cert.cert != nil &&
-			((*cert.cert).CheckSignatureFrom(cert.signingCert) == nil || cert.selfSign) {
+		// if regeneration is not needed - skip
+		if !regenerate(cs, oldCs, cert) {
+			fmt.Printf("skip %v\n", cert.cn)
 			continue
 		}
+		fmt.Printf("gen %v\n", cert.cn)
 		if *cert.key, *cert.cert, err = tls.NewCert(cert.cn, cert.organization, cert.dnsNames, cert.ipAddresses, cert.extKeyUsage, cert.signingKey, cert.signingCert, cert.selfSign); err != nil {
+			fmt.Println("bumot error")
```
Comment (jim-minter) [inline pkg/config/generate.go]: that's the worst kind of error.

File: pkg/config/generate.go [hunk 3]
```diff
@@ -388,3 +394,40 @@ func Generate(cs *api.OpenShiftManagedCluster) (err error) {
 
 	return
 }
+
+// regenerate checks if certificate needs to be regenerated
+func regenerate(cs, oldCs *api.OpenShiftManagedCluster, cert certificate) bool {
```
Comment (jim-minter) [inline pkg/config/generate.go]: this function needs to check the inputs against the pre-existing certificate in cs.Config to determine if a regeneration is needed.

File: pkg/config/generate.go (hunk 2 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: sorry, debug message :D

File: pkg/config/generate.go (hunk 3 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: This is where the challenge comes. My certificate might be for `old.example.com` but my `cs` is coming from `newawesomedomain.com`. So without `oldCs` im not able to check which certificate depends to where because values might be totally different. 
Alternatives are to keep some kind dependency between certificates by adding `type` to the certificate and mapping `type=masterFrontEnd` needs regeneration if `PublicHostname` or `FQDN` changes.

File: pkg/api/plugin.go (hunk 1 above)
Comment (mjudeikis) [reply pkg/api/plugin.go]: See last comment

File: pkg/config/generate.go (hunk 3 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: If you can think of any other way, let me know.

Comment (kargakis) [reply pkg/config/generate.go]: Is VerifyHostname what you need here?

https://golang.org/pkg/crypto/x509/#Certificate.VerifyHostname

Comment (mjudeikis) [reply pkg/config/generate.go]: The thing is I dont have a reference point to what to compare it. Without oldCS I need to know which cert uses which Properties field. In the example, routerPublicHostname is very similar to API FQDN, but they are not related. So verifyHostname will not give me this relationship.

Comment (kargakis) [reply pkg/config/generate.go]: At this point all the hostnames in cs are updated to new, so if a certificate does not match **any** hostname then you probably want to regenerate it, right?

Comment (mjudeikis) [reply pkg/config/generate.go]: This is just one case. In example:
```
 {
            cn: cs.Properties.RouterProfiles[0].PublicSubdomain,
            dnsNames: []string{
                cs.Properties.RouterProfiles[0].PublicSubdomain,
                "*." + cs.Properties.RouterProfiles[0].PublicSubdomain,
            },
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.Router.Key,
            cert:        &c.Certificates.Router.Cert,
        },
        {
            cn: "docker-registry-default." + cs.Properties.RouterProfiles[0].PublicSubdomain,
            dnsNames: []string{
                "docker-registry-default." + cs.Properties.RouterProfiles[0].PublicSubdomain,
                "docker-registry.default.svc",
                "docker-registry.default.svc.cluster.local",
            },
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.Registry.Key,
            cert:        &c.Certificates.Registry.Cert,
        },
        {
            cn: cs.Properties.FQDN,
            dnsNames: []string{
                cs.Properties.FQDN,
                "master-000000",
                "master-000001",
                "master-000002",
                "kubernetes",
                "kubernetes.default",
                "kubernetes.default.svc",
                "kubernetes.default.svc.cluster.local",
            },
            ipAddresses: []net.IP{net.ParseIP("172.30.0.1")},
            extKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
            key:         &c.Certificates.MasterServer.Key,
            cert:        &c.Certificates.MasterServer.Cert,
            
        }, 
```
certificates does not have purpose/type. Just fields If I change Router FQDN to totally different value, how do I know that I need to regenerate router certificate, and not Master or Service Catalog? They are not related, but without mapping keeping somewhere just matching is not enough.

Comment (kargakis) [reply pkg/config/generate.go]: Because master and service catalog are still going to match with the other hostname.

Comment (mjudeikis) [reply pkg/config/generate.go]: So you suggest to take app properties values, iterate over small subset of certificates (the ones which depend on those properties) and if none of them matches properties field, regenerate it. If match found - skip

Comment (kargakis) [reply pkg/config/generate.go]: I didn't suggest that but it's also an option to group certificates per "parent" hostname. I suggested that given a certificate, you can run VerifyHostname on it against all available hostnames (publicHostname, routerPublicHostname). If the cert cannot verify against any hostname, you want to regenerate it.

Comment (jim-minter) [reply pkg/config/generate.go]: @mjudeikis you need to write a function which compares the current certificate's fields to the cn, organization, dnsNames, ipAddresses, extKeyUsage desired state.  If they don't match, regenerate.  We already do this to determine if the CA has changed and we need to regenerate, so what's left is the easier bit?

**Comment by mjudeikis** [comment]
/hold
needs more testing as its sensitive change before full PP

**Comment by mjudeikis** [comment]
I would like to add more tests to this, but not sure how. Any ideas?

**Comment by mjudeikis** [comment]
@jim-minter @kargakis ptal

File: pkg/config/generate_test.go [hunk 4]
```diff
@@ -112,3 +113,54 @@ func testRequiredFields(cs *api.OpenShiftManagedCluster, t *testing.T) {
 	assert(c.NodeBootstrapKubeconfig != nil, "NodeBootstrapKubeconfig")
 	assert(c.AzureClusterReaderKubeconfig != nil, "AzureClusterReaderKubeconfig")
 }
+
+func TestGenerateUpdateCertRegen(t *testing.T) {
+	var pluginConfig api.PluginConfig
+	var cs, oldCs *api.OpenShiftManagedCluster
+	cs = fixtures.NewTestOpenShiftCluster()
+	cs.Properties.FQDN = "example-new.eastus.cloudapp.azure.com"
```
Comment (kargakis) [inline pkg/config/generate_test.go]: Can you move all of these where you start using cs? It's confusing to have them here.
In general I would try to factor down the loc
```go
	oldCs := fixtures.NewTestOpenShiftCluster()
	err := Generate(oldCs, api.PluginConfig{})
	if err != nil {
		t.Errorf("received generation error %v", err)
	}

	cs := fixtures.NewTestOpenShiftCluster()
	cs.Properties.FQDN = "example-new.eastus.cloudapp.azure.com"
	cs.Properties.RouterProfiles[0].FQDN = "router-fqdn-new.eastus.cloudapp.azure.com"
	cs.Properties.RouterProfiles[0].PublicSubdomain = "test-new.example.com"

	// copy config and regenerate everything else
	old := oldCs.DeepCopy()
	cs.Config = old.Config
	err = Generate(cs, pluginConfig)
	if err != nil {
		t.Errorf("received generation error %v", err)
	}

        // certificates should not match
	if reflect.DeepEqual(cs.Config.Certificates.MasterServer.Cert, oldCs.Config.Certificates.MasterServer.Cert) {
		t.Error("masterServer certificates matches, check test for details")
	}
	if reflect.DeepEqual(cs.Config.Certificates.Router.Cert, oldCs.Config.Certificates.Router.Cert) {
		t.Error("router certificates matches, check test for details")
	}
	if reflect.DeepEqual(cs.Config.Certificates.Registry.Cert, oldCs.Config.Certificates.Registry.Cert) {
		t.Error("registry certificates matches, check test for details")
	}
	if reflect.DeepEqual(cs.Config.Certificates.OpenshiftConsole.Cert, oldCs.Config.Certificates.OpenshiftConsole.Cert) {
		t.Error("openshiftConsole certificates matches, check test for details")
	}
	//certificates should match
	if reflect.DeepEqual(cs.Config.Certificates.Registry.Cert, oldCs.Config.Certificates.Registry.Cert) {
		t.Error("registry certificates matches, check test for details")
	}
	if !reflect.DeepEqual(cs.Config.Certificates.ServiceCatalogAPIClient.Cert, oldCs.Config.Certificates.ServiceCatalogAPIClient.Cert) {
		t.Error("serviceCatalogAPIClient certificates do not match, check test for details")
	}
	if !reflect.DeepEqual(cs.Config.Certificates.Admin.Cert, oldCs.Config.Certificates.Admin.Cert) {
		t.Error("admin certificates do not match, check test for details")
	}
}
```

File: pkg/config/generate.go [hunk 5]
```diff
@@ -227,13 +229,18 @@ func Generate(cs *api.OpenShiftManagedCluster, pluginConfig api.PluginConfig) (e
...
+		var k *rsa.PrivateKey
+		var c *x509.Certificate
+		if k, c, err = tls.NewCert(cert.cn, cert.organization, cert.dnsNames, cert.ipAddresses, cert.extKeyUsage, cert.signingKey, cert.signingCert, false); err != nil {
 			return
 		}
+		// compare certificate and replace if update is needed
+		// if at any point we start using self-sign certificates again
+		// (see console certificate comment), logic inside certEqual needs
+		// to be updated for this.
+		if !certEqual(c, *cert.cert) {
```
Comment (jim-minter) [inline pkg/config/generate.go]: This is the sole cpu-intensive part of the plugin.  It would be far better CPU wise to look at the old certificate and see if it is inconsistent with the new inputs, and only then generate a new certificate if needed.

File: pkg/config/generate.go [hunk 6]
```diff
@@ -388,3 +395,17 @@ func Generate(cs *api.OpenShiftManagedCluster, pluginConfig api.PluginConfig) (e
...
+
+func certEqual(certNew, certOld *x509.Certificate) bool {
+	if certOld == nil {
+		return false
+	}
+	if !reflect.DeepEqual(certNew.Subject, certOld.Subject) ||
+		!reflect.DeepEqual(certNew.DNSNames, certOld.DNSNames) ||
+		!reflect.DeepEqual(certNew.ExtKeyUsage, certOld.ExtKeyUsage) ||
+		!reflect.DeepEqual(certNew.IPAddresses, certOld.IPAddresses) ||
+		!reflect.DeepEqual(certNew.Issuer, certOld.Issuer) {
```
Comment (jim-minter) [inline pkg/config/generate.go]: Issuer is just a name.  This function will not fire correctly if the CA certificate is reissued and its name doesn't change.  That's why we use CheckSignatureFrom above.  You need to do the same.

File: pkg/config/generate.go [hunk 7]
```diff
@@ -15,6 +16,18 @@ import (
 	"github.com/openshift/openshift-azure/pkg/tls"
 )
 
+type certificate struct {
```
Comment (jim-minter) [inline pkg/config/generate.go]: unless you are actually really using this, revert this change

File: pkg/config/generate_test.go [hunk 8]
```diff
@@ -112,3 +113,54 @@ func testRequiredFields(cs *api.OpenShiftManagedCluster, t *testing.T) {
 	assert(c.NodeBootstrapKubeconfig != nil, "NodeBootstrapKubeconfig")
 	assert(c.AzureClusterReaderKubeconfig != nil, "AzureClusterReaderKubeconfig")
 }
+
+func TestGenerateUpdateCertRegen(t *testing.T) {
```
Comment (jim-minter) [inline pkg/config/generate_test.go]: This test will run pretty slowly.  Can't you just poke some pairs of certificates into certEqual and check that the return value is as expected?

File: pkg/config/generate.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: the challenge is that if lets say I change API FQDN. I have `config` object with Old values. And I have 18 certificates, without any distinguished mapping which certificate is using FQDN in any field. 
So when logic comes to `Router` certificate - It will not match any of the values, so logic would say that it needs to be regenerated. When in reality - it does not.

If we want to compare certificates with values without generation - certificates should have list of attributes assigned to them to know which certs should contain one or another attribute. 

I might be getting this from the wrong angle, but each time I dont generate certificate I get "mapping problem" to check right values with right certificates.

File: pkg/config/generate.go (hunk 7 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: It makes `regenerate` function cleaner. I can move it back but this means all logic from regenate goes to main function too...

File: pkg/config/generate_test.go (hunk 8 above)
Comment (mjudeikis) [reply pkg/config/generate_test.go]: go caches the test result, so not a big issue. And it shows how our other functions work in combined. I can do this, but I think this allows to test full flow too. WDYT?

File: pkg/config/generate.go [hunk 9]
```diff
@@ -388,3 +388,18 @@ func Generate(cs *api.OpenShiftManagedCluster, pluginConfig api.PluginConfig) (e
...
+	crt := *cert.cert
+	if crt == nil {
+		return false
+	}
+
+	if !reflect.DeepEqual(cert.dnsNames, crt.DNSNames) ||
+		!reflect.DeepEqual(cert.extKeyUsage, crt.ExtKeyUsage) ||
+		!reflect.DeepEqual(cert.cn, crt.Subject.CommonName) ||
+		!reflect.DeepEqual(cert.ipAddresses, crt.IPAddresses) {
+		return false
```
Comment (mjudeikis) [inline pkg/config/generate.go]: We might need to do some logic to check signing key too. but this is more complex.

File: pkg/config/generate.go [hunk 10]
```diff
@@ -388,3 +388,18 @@ func Generate(cs *api.OpenShiftManagedCluster, pluginConfig api.PluginConfig) (e
 
 	return
 }
+
+func regenerate(cert certificate) bool {
```
Comment (jim-minter) [inline pkg/config/generate.go]: call this needsGenerate and reverse the sense - ie where it's returning false currently return true, and vv

File: pkg/config/generate.go (hunk 9 above)
Comment (jim-minter) [reply pkg/config/generate.go]: also consider the organization

Comment (jim-minter) [reply pkg/config/generate.go]: the signing key shouldn't be too bad - if !selfSign, then call CheckSignatureFrom (the code you deleted above)

File: pkg/config/generate_test.go (hunk 8 above)
Comment (jim-minter) [reply pkg/config/generate_test.go]: I disagree.

File: pkg/config/generate.go (hunk 9 above)
Comment (mjudeikis) [reply pkg/config/generate.go]: ok, signing key was not hard. IP address was. Can you advise anything on this, please?

File: pkg/config/generate.go [hunk 11]
```diff
@@ -388,3 +391,43 @@ func Generate(cs *api.OpenShiftManagedCluster, pluginConfig api.PluginConfig) (e
...
+	}
+
+	// if main attributes do not match - regenerate
+	// net.IP does not comply with definition of deep equality.
+	// we use native Equal method within net package
+	if !reflect.DeepEqual(cert.dnsNames, crt.DNSNames) ||
+		!reflect.DeepEqual(cert.extKeyUsage, crt.ExtKeyUsage) ||
+		!reflect.DeepEqual(cert.cn, crt.Subject.CommonName) ||
+		!reflect.DeepEqual(cert.organization, crt.Subject.Organization) ||
+		!ipAddressEqual(cert.ipAddresses, crt.IPAddresses) {
```
Comment (mjudeikis) [inline pkg/config/generate.go]: can this be done better? basically net.IP return not equal all the times when using DeepEqual (similar topic https://stackoverflow.com/questions/42921227/ip-part-of-net-ipnet-does-not-fulfil-reflect-deepequal-but-are-equal) Can this be done better?

**Comment by jim-minter** [comment]
the functionality and testing here were incomplete.  I have completed and refactored into the tls package at  #561.
